### PR TITLE
Modifed code to run multiple pipelines in parallel and new environment 

### PR DESCRIPTION
**Pull request by mkaykeating**
- ipynb file to run multiple CP pipelines on one data set
- New environment to fix dependency issues

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# jupyter checkpoints
+*.ipynb_checkpoints
+
+# all images remain on local machine 
+**.tif
+
+# ingore intermediate
+**.npy
+
+**.log
```
Comment (MikeLippincott) [inline .gitignore:19]: Consider adding a comment here

File: Pipeline_example/nf1_ic.sh
Comment (MikeLippincott) [inline Pipeline_example/nf1_ic.sh:1]: Consider renaming this file name to better reflect the purpose?

File: Pipeline_example/scripts/cp_parallel.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
+
+
+"""
+Taken from Jenna's script from the Cellpainting repo
```
Comment (MikeLippincott) [inline Pipeline_example/scripts/cp_parallel.py]: Great string description of the script. Consider adding the link to the `cp_paralell.py` file

File: Pipeline_example/notebooks/multi_pipelines.ipynb
Comment (MikeLippincott) [inline Pipeline_example/notebooks/multi_pipelines.ipynb:1]: This notebook seems to be missing a python file. Consider adding the `.py` file

**Review by MikeLippincott — COMMENTED** [review]
This looks good, please see comments for considered changes. Also, consider updating the README file to include what the Pipeline example is? This could be done in a separate PR later if you choose to do this. Good work!

File: .gitignore (hunk 1 above)
Comment (mkaykeating) [reply .gitignore:19]: Done! Thanks!

File: Pipeline_example/nf1_ic.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Taken from Jenna's script from the Cellpainting repo
+
+#!/bin/bash
+
+# initialize the correct shell for your machine to allow conda to work (see README for note on shell names)
```
Comment (jenna-tomkinson) [inline Pipeline_example/nf1_ic.sh]:
```suggestion
# initialize the correct shell for your machine to allow conda to work
```
Recommend removing the section regarding the README unless you have a section on this added in your own README for this section.

File: Pipeline_example/scripts/cp_parallel.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+
+
+"""
```
Comment (jenna-tomkinson) [inline Pipeline_example/scripts/cp_parallel.py]: Recommend moving this docstring to the first line of the file.

File: Pipeline_example/scripts/cp_parallel.py (hunk 2 above)
Comment (jenna-tomkinson) [reply Pipeline_example/scripts/cp_parallel.py]: I agree with this recommendation! 😄

File: Pipeline_example/scripts/exceptions.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+"""
+Taken from Jenna's script from the Cellpainting repo
```
Comment (jenna-tomkinson) [inline Pipeline_example/scripts/exceptions.py]: Same recommendation from Mike above.

File: Pipeline_example/scripts/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+This class defines a custom exception class for exceeding the max workers on a machine.
+"""
+
+class MaxWorkerError(Exception):
+    """
+    Raised when the number of workers assigned to `max_workers` exceeds the number of CPU/workers on the machine. 
+    """
+    pass
```
Comment (jenna-tomkinson) [inline Pipeline_example/scripts/exceptions.py]: Add new endline to this file to remove error. Essentially what this means is the file is 10 lines of code but it needs one line at the bottom at is empty so there is a total of 11 lines.

File: Pipeline_example/nf1_ic.sh
Comment (mkaykeating) [reply Pipeline_example/nf1_ic.sh:1]: Renamed to run_ipynb_as_python_script. Thanks!

File: environments/cp_4.2.6.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+
+#modeled off of  "NF1_SchwannCell_data_analysis/environment.yml" & "nf1_cellpaining_data/nf1_cellpainting_env.yml"
```
Comment (jenna-tomkinson) [inline environments/cp_4.2.6.yml]: Recommend moving this environment code up one so that it starts at line 1.

File: Pipeline_example/scripts/cp_parallel.py (hunk 2 above)
Comment (mkaykeating) [reply Pipeline_example/scripts/cp_parallel.py]: Got it, thanks!

**Review by jenna-tomkinson — CHANGES_REQUESTED** [review]
Nice PR! Clean and small!

One thing you will need to do is add a python script of the `multi_pipelines.ipynb` so that we can review the file more easily.

Also, I would recommend changing the `scripts` folder that contains cp_parallel to `utils`, as we normally put notebook-converted files into a folder called either scripts or nbconverted.

I am putting this as requesting changes, but once you add the python script of the notebook, either Mike and I can review that and approve!

File: Pipeline_example/notebooks/multi_pipelines.ipynb
Comment (mkaykeating) [reply Pipeline_example/notebooks/multi_pipelines.ipynb:1]: Uploaded the .py file. thanks!

File: Pipeline_example/nf1_ic.sh (hunk 3 above)
Comment (mkaykeating) [reply Pipeline_example/nf1_ic.sh]: Got it, thank you!

File: Pipeline_example/scripts/cp_parallel.py (hunk 4 above)
Comment (mkaykeating) [reply Pipeline_example/scripts/cp_parallel.py]: Changed! Thanks!

File: Pipeline_example/scripts/exceptions.py (hunk 5 above)
Comment (mkaykeating) [reply Pipeline_example/scripts/exceptions.py]: Added a link in every doc. Thanks for the recommendation!

File: Pipeline_example/scripts/exceptions.py (hunk 6 above)
Comment (mkaykeating) [reply Pipeline_example/scripts/exceptions.py]: Got it, thanks!

File: environments/cp_4.2.6.yml (hunk 7 above)
Comment (mkaykeating) [reply environments/cp_4.2.6.yml]: Done!

**Comment by mkaykeating** [comment]
@MikeLippincott @jenna-tomkinson Hey! Thanks for the helpful comments, I really appreciate it! I was so excited my code was working I didn't focus on the organization so this helped me grasp what I can do better next time. I believe I made all of the changes ya'll suggested except updating the README doc. I will work on the README doc this afternoon. @jenna-tomkinson Should I add a "download_data" folder like you did in the cellpainting repo?

File: Pipeline_example/scripts/multi_pipelines.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+
+
+cp_parallel.run_cellprofiler_parallel(
+    plate_info_dictionary=plate_info_dictionary, run_name=run_name
+)
+
+
+
+
```
Comment (MikeLippincott) [inline Pipeline_example/scripts/multi_pipelines.py]: Consider removing the extra blank lines of code

File: Pipeline_example/utils/cp_parallel.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
+"""
+Taken from Jenna's Cellpainting repo (cp_parallel.py) https://github.com/WayScience/nf1_cellpainting_data/blob/main/utils/cp_parallel.py
```
Comment (MikeLippincott) [inline Pipeline_example/utils/cp_parallel.py:2]: Nice citation of the code origin, Love this!

File: Pipeline_example/utils/cp_parallel.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+This collection of functions runs CellProfiler in parallel and can convert the results into log files
```
Comment (MikeLippincott) [inline Pipeline_example/utils/cp_parallel.py:4]: Good documentation too

**Review by MikeLippincott — COMMENTED** [review]
LGTM, nice work!

**Comment by jenna-tomkinson** [comment]
@mkaykeating Great question! Don't worry about a download_data for this. I think that you can reference my repo for NF1 for how to download the datasets for reproducibility's sake.

**Comment by mkaykeating** [comment]
> @mkaykeating Great question! Don't worry about a download_data for this. I think that you can reference my repo for NF1 for how to download the datasets for reproducibility's sake.

@jenna-tomkinson Got it, thanks!

File: Pipeline_example/scripts/multi_pipelines.py
Comment (jenna-tomkinson) [inline Pipeline_example/scripts/multi_pipelines.py:1]: Looks like this file might not be nbconverted. I would recommend, even if you don't run the notebook fully to convert that file into a Python script directly.

This occurs when you run the `sh` file in terminal, so this might be resolved in a separate PR.

**Review by jenna-tomkinson — APPROVED** [review]
This PR looks wonderful! I left just one comment for you to address. Great job @mkaykeating!

File: Pipeline_example/scripts/multi_pipelines.py
Comment (mkaykeating) [reply Pipeline_example/scripts/multi_pipelines.py:1]: Thanks for looking through this so quickly! Can you please help me understand your comment? I converted the multi_pipelines.ipynb file to a python file just by saving it as a .py file. Are you saying I should use nbconvert  instead? Thanks!

Comment (jenna-tomkinson) [reply Pipeline_example/scripts/multi_pipelines.py:1]: Ahh okay so the difference is that when you just save as a Python file, it doesn't look like you are running code in multiple blocks like a notebook. It will look kinda like you are writing a function or utility file.

What I mean is that it will look like this file here: https://github.com/WayScience/nf1_cellpainting_data/blob/main/1.cellprofiler_ic/scripts/nf1_ic.py

You can see it has the `# In[1]` at the top. That means that the code under is run together before running the next block of code. 

You get this output when running Jupyter NBConvert like this:

```bash
# convert all notebooks to python files into the scripts folder
jupyter nbconvert --to python --output-dir=scripts/ *.ipynb
```

Let me know if you have any other questions! It might help to talk in person too, just let me know!

Comment (mkaykeating) [reply Pipeline_example/scripts/multi_pipelines.py:1]: That makes sense! I will work on converting them now, thanks!

File: Pipeline_example/scripts/multi_pipelines.py (hunk 8 above)
Comment (mkaykeating) [reply Pipeline_example/scripts/multi_pipelines.py]: Done, thanks!

File: Pipeline_example/scripts/multi_pipelines.py
Comment (mkaykeating) [inline Pipeline_example/scripts/multi_pipelines.py:1]: @jenna-tomkinson Here is my nbconverted file. I believe I have addressed everything else. Am I all good to merge? Thanks!
